### PR TITLE
#167750054 Implement social authentication using Google and Facebook

### PR DESCRIPTION
**Pull request by darasimiolaifa**
**What does this PR do?**
This PR implements login functionality using social authentication through platforms like Google and Facebook

**Description of tasks to be completed?**
- create Google app in Google developer console
- create Facebook app in Facebook developer console
- install passport packages to talk to each social platform
- create endpoints to allow users use the feature
- write passport middleware and controllers to handle the routes

**How should this be manually tested/checked?**
1. In your terminal, run `git clone https://github.com/andela/wolfsbane-backend.git`
2. Run `cd wolfsbane-backend`
3. Run `git fetch origin ft-social-authentication-167750054`
4. Run `npm install`
5. Run `npm run migrate` to migrate the model definitions to both databases (`development` and `test`)
6. Setup .env file with the following keys `GOOGLE_CLIENT_ID`, `GOOGLE_CLIENT_SECRET`, `FACEBOOK_CLIENT_ID`, `FACEBOOK_CLIENT_SECRET`, `SECRET`, and their respective values as shared on the team slack channel. Also set `PASSWORD` to a string of characters.
7. Run `npm run start:dev` to start the app
8. For Facebook, in your browser, open tab in private/incognito and visit `http://localhost:3000/api/v1/auth/facebook`
9. Enter the login credentials as shared on the slack channel
10. You should be redirected to a page that returns an object that contains a `status code` of `200` and a `token` string.
11. For Google, in your browser, visit `http://localhost:3000/api/v1/auth/google`
12. Login with your email and password.
13. You should be redirected to a similar page like the Facebook's case.
14. In both cases, there should be entries in your `development` database that correspond to your credentials on the respective platforms.
16. To run automated tests, run `npm run test`.

**Any background context you want to provide?**
- You need the incognito mode in Facebook's case to prevent Facebook from trying to use an already logged in user to access the app, as this won't work because only registered test accounts can access the app.
- To test in the browser (using `npm run start:dev`), make sure your `dev` databse is set up, migrated to, and is running. While to test using automation (`npm run test`), make sure your `test` database is set up, migrated to, and running.

**What are the relevant pivotal tracker stories?**
[#167750054](https://www.pivotaltracker.com/story/show/167750054)

### CODE REVIEW

**Review by feobaby — CHANGES_REQUESTED** [review]
Great work here Darasimi, I was trying to test it and i realized that a table `Users` does not exist, please change that sequelize migration table to `users` 

this is the error ` UnhandledPromiseRejectionWarning: Error: Error: secretOrPrivateKey must have a value` when logging in with FB

**Comment by darasimiolaifa** [comment]
> Great work here Darasimi, I was trying to test it and i realized that a table `Users` does not exist, please change that sequelize migration table to `users`
> 
> this is the error ` UnhandledPromiseRejectionWarning: Error: Error: secretOrPrivateKey must have a value` when logging in with FB

Thanks @funmi5. Can you do me a favor and check what the URL is in your browser tab where you get the error? Plus did you set the APP_SECRET in your .env file?

**Comment by feobaby** [comment]
> > Great work here Darasimi, I was trying to test it and i realized that a table `Users` does not exist, please change that sequelize migration table to `users`
> > this is the error ` UnhandledPromiseRejectionWarning: Error: Error: secretOrPrivateKey must have a value` when logging in with FB
> 
> Thanks @funmi5. Can you do me a favor and check what the URL is in your browser tab where you get the error? Plus did you set the APP_SECRET in your .env file?

Hm, no I didn't . i dont think you supplied us with it.

**Comment by darasimiolaifa** [comment]
> > > Great work here Darasimi, I was trying to test it and i realized that a table `Users` does not exist, please change that sequelize migration table to `users`
> > > this is the error ` UnhandledPromiseRejectionWarning: Error: Error: secretOrPrivateKey must have a value` when logging in with FB
> > 
> > 
> > Thanks @funmi5. Can you do me a favor and check what the URL is in your browser tab where you get the error? Plus did you set the APP_SECRET in your .env file?
> 
> Hm, no I didn't . i dont think you supplied us with it.

You can set it to any value. Any value works. APP_SECRET is peculiar to any instantiation of the app. Whatever value you set will work.

**Review by oyewoas — CHANGES_REQUESTED** [review]
well done @darasimiolaifa, the app works but i am getting an error when I run tests  for **Social Authentication Test**  the error goes thus:
`Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.`

**Comment by sa-ma** [comment]
I think you should structure your work. You have so much going on in your routes file. You should create a controller or even a service to handle the business logic

**Comment by darasimiolaifa** [comment]
> I think you should structure your work. You have so much going on in your routes file. You should create a controller or even a service to handle the business logic

Ok thanks. I really should, since the finished product will be merged into the codebase, but one reason why it is like that is because some (not all though) of those functions are temporary ones that I need from others. But like you suggested, I'll refactor mine, and press and help them get theirs approved. Currently walking through reviewing @oyewoas's code. Thanks.

**Comment by darasimiolaifa** [comment]
> well done @darasimiolaifa, the app works but i am getting an error when I run tests for **Social Authentication Test** the error goes thus:
> `Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.`

I think I fixed this. (Travis isn't building, so I am not sure). Can you please review and confirm? Thanks.

**Comment by darasimiolaifa** [comment]
> > > > Great work here Darasimi, I was trying to test it and i realized that a table `Users` does not exist, please change that sequelize migration table to `users`
> > > > this is the error ` UnhandledPromiseRejectionWarning: Error: Error: secretOrPrivateKey must have a value` when logging in with FB
> > > 
> > > 
> > > Thanks @funmi5. Can you do me a favor and check what the URL is in your browser tab where you get the error? Plus did you set the APP_SECRET in your .env file?
> > 
> > 
> > Hm, no I didn't . i dont think you supplied us with it.
> 
> You can set it to any value. Any value works. APP_SECRET is peculiar to any instantiation of the app. Whatever value you set will work.

Can you please review and confirm the effecting of the changes you requested? Thanks. Don't forget to set `APP_SECRET` in your .env file. Any value will do. Thanks.

**Comment by darasimiolaifa** [comment]
> I think you should structure your work. You have so much going on in your routes file. You should create a controller or even a service to handle the business logic

Done. Can you please confirm?

Also, can you give me a detailed account of how to reproduce the bug you described on the chat? I tried several combinations, and in each case, Facebook responded with their own customized error that prevented the situation you described.

Thanks.

**Comment by sa-ma** [comment]
> > I think you should structure your work. You have so much going on in your routes file. You should create a controller or even a service to handle the business logic
> 
> Done. Can you please confirm?
> 
> Also, can you give me a detailed account of how to reproduce the bug you described on the chat? I tried several combinations, and in each case, Facebook responded with their own customized error that prevented the situation you described.
> 
> Thanks.

The issue came from the fact that I was already signed in which you have addressed  also all the tests are failing

File: src/controllers/authController.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import passport from 'passport';
+import JWT from 'jsonwebtoken';
+import { config } from 'dotenv';
+import passportConfIgnored from '../config/passport';
```
Comment (sa-ma) [inline src/controllers/authController.js:3]: This package wasn't used

**Comment by darasimiolaifa** [comment]
Okay. Can you please help with the error message? Please, not the timeout error, but the real error that follows each of the test suites. Or better still, send a screenshot to the slack channel. Thanks.

File: src/controllers/authController.js (hunk 1 above)
Comment (darasimiolaifa) [reply src/controllers/authController.js:3]: It isn't a package. It's a configuration, and weirdly enough, passport won't work without it being imported this way. No other reference to it, just import it into the scope here.

**Review by oyewoas — DISMISSED** [review]

**Review by Oliver-ke — DISMISSED** [review]
Nice job @darasimiolaifa, though more things  regarding authentication would be added later

**Review by feobaby — DISMISSED** [review]
Really really nice work, keep it up @darasimiolaifa

File: src/config/mockStrategy.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
+/* eslint-disable import/no-extraneous-dependencies */
```
Comment (eokwukwe) [inline src/config/mockStrategy.js]: Why is this added?

File: src/controllers/authController.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    passport.authenticate(
+      provider,
+      (err, user) => {
+        if (!user) {
+          return res.status(401).send({ status: 401, message: 'You\'re not authenticated' });
+        }
+        const token = JWT.sign({
+          iss: 'Wolfsbane-BarefootNomad',
+          sub: user,
+        }, process.env.APP_SECRET);
```
Comment (eokwukwe) [inline src/controllers/authController.js]: Token creation and verification should be abstracted to a function in the `utils` that can be imported and used anywhere.

File: src/utils/passportStrategyCallback.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import database from '../models';
```
Comment (eokwukwe) [inline src/utils/passportStrategyCallback.js]: use `models` instead of `database` for consistency

File: src/config/mockStrategy.js (hunk 2 above)
Comment (darasimiolaifa) [reply src/config/mockStrategy.js]: It is to prevent eslint from throwing errors, because I am bringing in two packages (`passport-strategy` and `util`) that are in my `node_modules` but not in my `dependencies` or `devDependencies`, because they are installed by packages that I use. For instance, `passport-strategy` is installed by `passport`, and I think `util` is part of the `express` framework. But eslint will not permit me to use them unless they are listed in my dependencies.

File: src/controllers/authController.js (hunk 3 above)
Comment (darasimiolaifa) [reply src/controllers/authController.js]: I used it here like this because I didn't want to duplicate methods. @oyewoas  is already writing that method, but his work hasn't been merged yet.

**Review by eokwukwe — COMMENTED** [review]
@darasimiolaifa good job so far. However, your work lacks proper documentation. Please ensure to add appropriate documentation to your work.

File: src/utils/passportStrategyCallback.js (hunk 4 above)
Comment (darasimiolaifa) [reply src/utils/passportStrategyCallback.js]: Okay. On it.

File: src/config/mockStrategy.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+import passport from 'passport-strategy';
+import util from 'util';
```
Comment (eokwukwe) [inline src/config/mockStrategy.js:2]: What is `util` used for?

Comment (darasimiolaifa) [reply src/config/mockStrategy.js:2]: It is used in `line 19`:
 `util.inherits(Strategy, passport.Strategy);`.
It is making `Strategy` inherit `passport.Strategy`, by setting the prototype of `Strategy` to an instance of the prototype of `passport.Strategy`.

That way, we can pass `Strategy` as a middleware to be used by `passport` and override the `authenticate` method that `Strategy` inherits with our own custom `authenticate` method.

**Comment by sa-ma** [comment]
Please, edit your PR description.
APP_SECRET env variable didn't work for me 
Also, all we need to do is run `npm run migrate` and `npm run seed` to set up the database instead of manually checking if the required fields are there
You should add a catch block that reports network error if the network call fails
And work on your test coverage for passportStrategyCallback.js in the utils folder

**Comment by darasimiolaifa** [comment]
> Please, edit your PR description.
> APP_SECRET env variable didn't work for me
> Also, all we need to do is run `npm run migrate` and `npm run seed` to set up the database instead of manually checking if the required fields are there
> You should add a catch block that reports network error if the network call fails
> And work on your test coverage for passportStrategyCallback.js in the utils folder

Alright, on it. Thanks.

**Comment by darasimiolaifa** [comment]
> > Please, edit your PR description.
> > APP_SECRET env variable didn't work for me
> > Also, all we need to do is run `npm run migrate` and `npm run seed` to set up the database instead of manually checking if the required fields are there
> > You should add a catch block that reports network error if the network call fails
> > And work on your test coverage for passportStrategyCallback.js in the utils folder
> 
> Alright, on it. Thanks.

By `adding a catch block that reports a network error if network call fails`, are you talking about the call to the database? Can you help me point to where in my code you're referring to? Thanks.

**Comment by darasimiolaifa** [comment]
> > > Please, edit your PR description.
> > > APP_SECRET env variable didn't work for me
> > > Also, all we need to do is run `npm run migrate` and `npm run seed` to set up the database instead of manually checking if the required fields are there
> > > You should add a catch block that reports network error if the network call fails
> > > And work on your test coverage for passportStrategyCallback.js in the utils folder
> > 
> > 
> > Alright, on it. Thanks.
> 
> By `adding a catch block that reports a network error if network call fails`, are you talking about the call to the database? Can you help me point to where in my code you're referring to? Thanks.

Done.

File: src/config/passport.js [hunk 6]
```diff
@@ -1,27 +1,30 @@
-const passport = require('passport');
-const LocalStrategy = require('passport-local').Strategy;
-const mongoose = require('mongoose');
+import passport from 'passport';
+import { OAuth2Strategy as GoogleStrategy } from 'passport-google-oauth';
+import { Strategy as FacebookStrategy } from 'passport-facebook';
+import MockStrategy from './mockStrategy';
```
Comment (eokwukwe) [inline src/config/passport.js:4]: Why do we need a mock strategy?

File: src/models/User.js [hunk 7]
```diff
@@ -16,6 +16,9 @@ export default (sequelize, DataTypes) => {
     password: {
       type: DataTypes.STRING,
       allowNull: false
+    },
+    isVerified: {
+      type: DataTypes.BOOLEAN,
```
Comment (eokwukwe) [inline src/models/User.js]: Why is the default value missing in the model?

File: src/config/passport.js (hunk 6 above)
Comment (darasimiolaifa) [reply src/config/passport.js:4]: It is how we are testing the code written.

That way I can tell passport to use this strategy when we are in a test environment. So my test isn't mocking any function. It is using the actual passport library to test,

That way I can use the same url I use for production, and test the workflow like I am really authenticating through those platforms. It is the reason why our mock strategy is inheriting from `passport.Strategy`.

File: src/models/User.js (hunk 7 above)
Comment (darasimiolaifa) [reply src/models/User.js]: Okay. I will insert it now.

File: src/controllers/authController.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  const callbackFunction = (req, res, next) => {
+    if (req.headers.testunauthorized) provider = 'unauthorized';
+    passport.authenticate(
+      provider,
+      (err, user) => {
+        if (!user) {
+          return res.status(401).send({ status: 401, message: 'You\'re not authenticated' });
+        }
+        const token = generateUserToken(user);
+        res.send({ status: res.statusCode, token });
```
Comment (eokwukwe) [inline src/controllers/authController.js]: The response should be a `JSON`.

File: src/controllers/authController.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+   * @returns {function} callbackFunction
+*/
+export const callback = (provider) => {
+  const callbackFunction = (req, res, next) => {
+    if (req.headers.testunauthorized) provider = 'unauthorized';
+    passport.authenticate(
+      provider,
+      (err, user) => {
+        if (!user) {
+          return res.status(401).send({ status: 401, message: 'You\'re not authenticated' });
```
Comment (eokwukwe) [inline src/controllers/authController.js]: Your returned response should be in JSON

File: src/config/passport.js (hunk 6 above)
Comment (darasimiolaifa) [reply src/config/passport.js:4]: So, the strategy function handles the dual cases of failed authentication (where no user object is returned), and a successful one (where a user object is returned).

File: src/models/User.js (hunk 7 above)
Comment (eokwukwe) [reply src/models/User.js]: The default value is still missing

File: src/test/models/Profile.test.js [hunk 10]
```diff
@@ -14,22 +14,22 @@ describe('Test for Profile Model', () => {
   const Profile = ProfileModel(sequelize, dataTypes);
   const profiles = new Profile();
 
-  checkModelName(Profile)('profiles');
+  checkModelName(Profile)('Profiles');
 
   context('properties', () => {
-    ['userId', 'phoneNumber', 'departmentId', 'role', 'jobDescription', 'imageUrl', 'isVerified'].forEach(
+    ['userId', 'phoneNumber', 'departmentId', 'role', 'jobDescription', 'imageUrl'].forEach(
```
[the remark this answers is not in the document]
Comment (darasimiolaifa) [reply src/test/models/Profile.test.js]: @eokwukwe `eslint` will throw an error here if I do what `hound` is asking me to do. Should I disable `eslint` for the line, or ignore this? Plus I think following `eslint` is cleaner here than listening to `hound`.

File: src/controllers/authController.js [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+   * @returns {function} callbackFunction
+*/
+export const callback = provider => (req, res, next) => {
+  passport.authenticate(
+    provider,
+    (err, user) => {
+      // check if server returns an error
+      if (err) return res.status(500).json({ status: 500, message: 'Registration unsuccessful. Please try again later.' });
+      // check if user authentication failed on social platform
+      // eslint-disable-next-line no-else-return
```
Comment (darasimiolaifa) [inline src/controllers/authController.js]: @eokwukwe I had to use inline rule here because `eslint` thinks I should not write an `else` statement after a `return` statement (which I ordinarily won't do too), but if I take `eslint`'s advice and use two different `if` statements, it increases the `cognitive complexity`, and `codeclimate` complains. Yet I have to check for both conditions and return different `status codes` and `messages` to the user, and I don't want to set an application-wide rule in the `.eslintrc` file

Comment (eokwukwe) [reply src/controllers/authController.js]: You don't need the `else if` block because the callback will return either `error` or `user`. The check for `err` is the same as the check for `!user`. Hence all you need to check is for error. The error will be returned if the user uses an invalid credential for the third party authentication. Your code should look like this:

```
if (err) return res.status(500).json({ status: 500, message: 'Registration unsuccessful. Please try again later.' });

// process successfully authenticated and registered user
const token = generateUserToken(user);
res.status(res.statusCode).json({ status: res.statusCode, token });
```

or if you still want to check for the user, you may write it this way:
```
if (err || !user) return res.status(500).json({ status: 500, message: 'Registration unsuccessful. Please try again later.' });

// process successfully authenticated and registered user
const token = generateUserToken(user);
res.status(res.statusCode).json({ status: res.statusCode, token });
```
but the first is enough and preferred

File: src/utils/passportStrategyCallback.js [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+const strategyCallback = async (accessToken, refreshToken, profile, done) => {
+  try {
+    if (!profile) return done(null, undefined);
+    const email = profile.emails[0].value;
+    const existingUser = await models.Users.findOne({ where: { email } });
+    if (existingUser) {
+      return done(null, existingUser);
+    }
+    const { name: { familyName: lastName, givenName: firstName } } = profile;
+    const password = Math.random().toString(36).slice(2);
```
Comment (eokwukwe) [inline src/utils/passportStrategyCallback.js]: Create a password in the .env file and use the password here. Remember to hash the password. This is to maintain consistency.

File: src/controllers/authController.js (hunk 11 above)
Comment (darasimiolaifa) [reply src/controllers/authController.js]: Can I use the second? Because in my test, to increase the coverage, I am actually testing for both scenarios differently. The error test is detecting error from the database server, while the null user is detecting error from authentication from the social media platform..

Comment (eokwukwe) [reply src/controllers/authController.js]: Yes you can

File: src/utils/passportStrategyCallback.js (hunk 12 above)
Comment (darasimiolaifa) [reply src/utils/passportStrategyCallback.js]: Done.

File: src/controllers/authController.js (hunk 11 above)
Comment (darasimiolaifa) [reply src/controllers/authController.js]: Done.

**Review by eokwukwe — COMMENTED** [review]
You need to update your PR with the new detail about the password in the `.env` file

File: src/utils/passportStrategyCallback.js [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+const strategyCallback = async (accessToken, refreshToken, profile, done) => {
+  try {
+    if (!profile) return done(null, undefined);
+    const email = profile.emails[0].value;
+    const existingUser = await models.Users.findOne({ where: { email } });
+    if (existingUser) {
+      return done(null, existingUser);
+    }
+    const { name: { familyName: lastName, givenName: firstName } } = profile;
+    const password = config.password;
```
Comment (eokwukwe) [inline src/utils/passportStrategyCallback.js]: You need to hash the password before saving it to the database

Comment (darasimiolaifa) [reply src/utils/passportStrategyCallback.js]: Done.

**Comment by darasimiolaifa** [comment]
> You need to update your PR with the new detail about the password in the `.env` file

I have done that. Thanks for pointing it out.
